packages/node/src/indexer/unfinalizedBlocks.service.ts

```diff
@@ -0,0 +1,239 @@
+  private removeFinalized(blockHeight: number): void {
+    Object.entries(this.unfinalizedBlocks).map(([bestBlockHeight, hash]) => {
+      if (Number(bestBlockHeight) <= blockHeight) {
+        delete this.unfinalizedBlocks[bestBlockHeight];
+      }
+    });
+  }
+
+  // find closest record from block heights
+  private getClosestRecord(
+    blockHeight: number,
+  ): { blockHeight: number; hash: HexString } | undefined {
+    // Have the block in the best block, can be verified
+    const record = this.getSortedUnfinalizedBlocks().find(
+      ([bestBlockHeight, hash]) => Number(bestBlockHeight) <= blockHeight,
+    );
+    if (record) {
+      const [bestBlockHeight, hash] = record;
+      return { blockHeight: Number(bestBlockHeight), hash };
+    }
+    return undefined;
+  }
+
+  // check unfinalized blocks for a fork
+  private async hasForked(): Promise<boolean> {
+    const lastVerifiableBlock = this.getClosestRecord(
+      this.finalizedBlockNumber,
+    );
+
+    // No unfinalized blocks
+    if (!lastVerifiableBlock) {
+      return false;
+    }
+
+    // Unfinalized blocks beyond finalized block
+    if (lastVerifiableBlock.blockHeight === this.finalizedBlockNumber) {
+      return lastVerifiableBlock.hash !== this.finalizedHeader.hash.toHex();
+    } else {
+      // Unfinalized blocks below finalized block
+      let header = this.finalizedHeader;
+      /*
+       * Iterate back through parent hashes until we get the header with the matching height
+       * We use headers here rather than getBlockHash because of potential caching issues on the rpc
+       * If we're off by a large number of blocks we can optimise by getting the block hash directly
+       */
+      if (header.number.toNumber() - lastVerifiableBlock.blockHeight > 200) {
+        const hash = await this.api.rpc.chain.getBlockHash(
+          lastVerifiableBlock.blockHeight,
+        );
+        header = await this.api.rpc.chain.getHeader(hash);
+      } else {
+        while (lastVerifiableBlock.blockHeight !== header.number.toNumber()) {
+          header = await this.api.rpc.chain.getHeader(header.parentHash);
+        }
+      }
+      if (header.hash.toHex() !== lastVerifiableBlock.hash) {
+        logger.warn(
+          `Block fork found, enqueued un-finalized block at ${
+            lastVerifiableBlock.blockHeight
+          } with hash ${
+            lastVerifiableBlock.hash
+          }, actual hash is ${header.hash.toHex()} `,
+        );
+        return true;
+      }
+    }
+  }
+
+  private async getLastCorrectFinalizedBlock(): Promise<number | undefined> {
+    const bestVerifiableBlocks = this.getSortedUnfinalizedBlocks().filter(
+      ([bestBlockHeight, hash]) =>
+        Number(bestBlockHeight) <= this.finalizedBlockNumber,
+    );
+
+    let checkingHeader = this.finalizedHeader;
+
+    // Work backwards through the blocks until we find a matching hash
+    for (const [block, hash] of bestVerifiableBlocks.reverse()) {
```

Comment:
this is still somehow duplicated with `hasForked()`.
Also, as I pointed out, `bestVerifiableBlocks` is not continuous, so you can't call `checkingHeader` only once in a loop.

Reply:
`bestVerifiableBlocks` should be continuous as we do it for every block we index and the dictionaries should not be indexing unfinalized block so it will be disabled